core/types/access_witness.go

```diff
@@ -17,94 +17,168 @@
 package types
 
 import (
+	"fmt"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/params"
 )
 
+type VerkleStem [31]byte
+
+type ChunkValue struct {
+	mode byte
+	value []byte
+}
+
 // AccessWitness lists the locations of the state that are being accessed
 // during the production of a block.
 // TODO(@gballet) this doesn't fully support deletions
 type AccessWitness struct {
 	// Branches flags if a given branch has been loaded
-	Branches map[[31]byte]struct{}
+	// for the byte value:
+	//	the first bit is set if the branch has been edited
+	//	the second bit is set if the branch has been read
+	Branches map[VerkleStem]byte
 
 	// Chunks contains the initial value of each address
-	Chunks map[common.Hash][]byte
-
-	// The initial value isn't always available at the time an
-	// address is touched, this map references addresses that
-	// were touched but can not yet be put in Chunks.
-	Undefined map[common.Hash]struct{}
+	Chunks map[common.Hash]ChunkValue
 }
 
 func NewAccessWitness() *AccessWitness {
 	return &AccessWitness{
-		Branches:  make(map[[31]byte]struct{}),
-		Chunks:    make(map[common.Hash][]byte),
-		Undefined: make(map[common.Hash]struct{}),
+		Branches:  make(map[VerkleStem]byte),
+		Chunks:    make(map[common.Hash]ChunkValue),
+	}
+}
+
+const (
+	AccessWitnessReadFlag = 1
+	AccessWitnessWriteFlag = 2
+)
```

Comment:
Move this above `ChunkState`, declare a `Mode` type and set it like:

```suggestion
const (
	AccessWitnessReadFlag = Mode(1)
	AccessWitnessWriteFlag = Mode(2)
)
```